src/librustc/ty/context.rs

```diff
@@ -1089,6 +1094,52 @@ impl<'tcx> TyCtxt<'tcx> {
         &self.hir_map
     }
 
+    /// In interpreter mode, gets the `DefId` of the user fn (a closure marked by an attribute named
+    /// `rustc_interp_user_fn`). This first time this method is called, that fn is located in the
+    /// crate HIR, and is cached for subsequent calls.
+    #[inline]
+    pub fn get_interp_user_fn(&self) -> DefId {
+        use hir::intravisit::{self, Visitor, NestedVisitorMap};
+
+        struct InterpUserFnVisitor<'tcx> {
+            tcx: TyCtxt<'tcx>,
+            def_id: Option<DefId>,
+        }
+
+        impl<'tcx> Visitor<'tcx> for InterpUserFnVisitor<'tcx> {
+            fn nested_visit_map<'this>(&'this mut self) -> NestedVisitorMap<'this, 'tcx> {
+                NestedVisitorMap::OnlyBodies(&self.tcx.hir())
+            }
+
+            fn visit_expr(&mut self, ex: &'tcx hir::Expr) {
+                if syntax::attr::contains_name(&ex.attrs, sym::rustc_interp_user_fn) {
+                    self.def_id = Some(self.tcx.hir().local_def_id(ex.hir_id));
+                    return;
+                }
+
+                intravisit::walk_expr(self, ex);
+            }
+        }
+
+        if let Some(&def_id) = self.interp_user_fn.try_get() {
+            def_id
+        } else {
+            let mut visitor = InterpUserFnVisitor {
+                tcx: *self,
+                def_id: None,
+            };
+            self.hir().krate().visit_all_item_likes(&mut visitor.as_deep_visitor());
+            let def_id = visitor.def_id
+                .expect(&format!(
```

Comment:
I think it'd be better to do something like `self.sess.err(...)` rather than expect here, this'll lead to a common ICE.

Reply:
Okay.